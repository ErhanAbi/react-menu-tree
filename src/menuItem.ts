export interface MenuItem {
  title: string;
  url: string;
  children?: MenuItem[];
}

export const filterMenuItems = (menuItems: MenuItem[], keyword: string) => {
  if (!keyword) {
    return menuItems;
  }

  const filtered: MenuItem[] = menuItems.filter(menuItem => {
    const containsKeyword =
      menuItem.title.toLowerCase().indexOf(keyword.toLowerCase()) > -1;
    const childContainsKeyword: boolean = !!menuItem.children
      ? filterMenuItems(menuItem.children, keyword).length > 0
      : false;

    return containsKeyword || childContainsKeyword;
  });

  return filtered;
};
