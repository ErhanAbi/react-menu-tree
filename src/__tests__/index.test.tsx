import React from "react";
import renderer from "react-test-renderer";

import { Menu } from "..";
import { MenuItem } from "../menuItem";

const testMenuList: MenuItem[] = [
  {
    title: "Getting started",
    url: "getting-started",
    children: [
      { title: "First", url: "getting-started/first" },
      { title: "Installation", url: "getting-started/installation" },
      { title: "Site preparation", url: "getting-started/site-preparation" },
      {
        title: "Creating your site",
        url: "getting-started/creating-your-site"
      },
      {
        title: "Publishing your site",
        url: "getting-started/publishing-your-site"
      }
    ]
  },
  {
    title: "Guides",
    url: "guides",
    children: [
      {
        title: "Adding a Blog",
        url: "guides/adding-a-blog",
        children: [{ title: "First", url: "guides/adding-a-blog/first" }]
      },
      {
        title: "Custom Pages",
        url: "guides/custom-pages",
        children: [{ title: "First", url: "guides/custom-pages/first" }]
      },
      { title: "Enabling Search", url: "guides/enabling-search" },
      {
        title: "Navigation and Sidebars",
        url: "guides/navigation-and-sidebars"
      },
      {
        title: "Translations & Localization",
        url: "guides/trans-and-nav"
      },
      { title: "Versioning", url: "guides/versioning" }
    ]
  }
];

describe("Menu", () => {
  it("renders menu based on the given list of menu items", () => {
    const wrapper = renderer
      .create(<Menu options={testMenuList} highlightClass="highlight" />)
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders menu based on the given list and highlights the selected menu item", () => {
    const wrapper = renderer
      .create(
        <Menu
          options={testMenuList}
          selectedOption="/guides/versioning"
          highlightClass="highlight"
        />
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
