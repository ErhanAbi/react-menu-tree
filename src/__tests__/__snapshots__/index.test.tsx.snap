// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders menu based on the given list and highlights the selected menu item 1`] = `
<div
  className={undefined}
>
  <input
    className={undefined}
    onChange={[Function]}
    placeholder="Search menu..."
    type="text"
    value=""
  />
  <ul
    className={undefined}
  >
    <li>
      <a
        className=""
        href="/getting-started"
      >
        Getting started
      </a>
      <ul
        className={undefined}
      >
        <li>
          <a
            className=""
            href="/getting-started/first"
          >
            First
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/installation"
          >
            Installation
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/site-preparation"
          >
            Site preparation
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/creating-your-site"
          >
            Creating your site
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/publishing-your-site"
          >
            Publishing your site
          </a>
        </li>
      </ul>
    </li>
    <li>
      <a
        className=""
        href="/guides"
      >
        Guides
      </a>
      <ul
        className={undefined}
      >
        <li>
          <a
            className=""
            href="/guides/adding-a-blog"
          >
            Adding a Blog
          </a>
          <ul
            className={undefined}
          >
            <li>
              <a
                className=""
                href="/guides/adding-a-blog/first"
              >
                First
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            className=""
            href="/guides/custom-pages"
          >
            Custom Pages
          </a>
          <ul
            className={undefined}
          >
            <li>
              <a
                className=""
                href="/guides/custom-pages/first"
              >
                First
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            className=""
            href="/guides/enabling-search"
          >
            Enabling Search
          </a>
        </li>
        <li>
          <a
            className=""
            href="/guides/navigation-and-sidebars"
          >
            Navigation and Sidebars
          </a>
        </li>
        <li>
          <a
            className=""
            href="/guides/trans-and-nav"
          >
            Translations & Localization
          </a>
        </li>
        <li>
          <a
            className={undefined}
            href="/guides/versioning"
          >
            Versioning
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`Menu renders menu based on the given list of menu items 1`] = `
<div
  className={undefined}
>
  <input
    className={undefined}
    onChange={[Function]}
    placeholder="Search menu..."
    type="text"
    value=""
  />
  <ul
    className={undefined}
  >
    <li>
      <a
        className=""
        href="/getting-started"
      >
        Getting started
      </a>
      <ul
        className={undefined}
      >
        <li>
          <a
            className=""
            href="/getting-started/first"
          >
            First
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/installation"
          >
            Installation
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/site-preparation"
          >
            Site preparation
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/creating-your-site"
          >
            Creating your site
          </a>
        </li>
        <li>
          <a
            className=""
            href="/getting-started/publishing-your-site"
          >
            Publishing your site
          </a>
        </li>
      </ul>
    </li>
    <li>
      <a
        className=""
        href="/guides"
      >
        Guides
      </a>
      <ul
        className={undefined}
      >
        <li>
          <a
            className=""
            href="/guides/adding-a-blog"
          >
            Adding a Blog
          </a>
          <ul
            className={undefined}
          >
            <li>
              <a
                className=""
                href="/guides/adding-a-blog/first"
              >
                First
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            className=""
            href="/guides/custom-pages"
          >
            Custom Pages
          </a>
          <ul
            className={undefined}
          >
            <li>
              <a
                className=""
                href="/guides/custom-pages/first"
              >
                First
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            className=""
            href="/guides/enabling-search"
          >
            Enabling Search
          </a>
        </li>
        <li>
          <a
            className=""
            href="/guides/navigation-and-sidebars"
          >
            Navigation and Sidebars
          </a>
        </li>
        <li>
          <a
            className=""
            href="/guides/trans-and-nav"
          >
            Translations & Localization
          </a>
        </li>
        <li>
          <a
            className=""
            href="/guides/versioning"
          >
            Versioning
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
